lib/UR/Observer.pm

```diff
@@ -40,68 +41,111 @@ sub __define__ {
+        }
+        $inner_self->{callback} = $callback;
+        $inner_self->register_callback(map { $_ => $inner_self->$_ } @required_params_for_register);
+        return $inner_self;
+    };
 
-    my ($rule,%extra) = UR::BoolExpr->resolve($class,@_);
-    my $callback = delete $extra{callback};
-    unless ($callback) {
-        $class->error_message("'callback' is a required parameter for creating UR::Observer objects");
-        return;
-    }
-    if (%extra) {
-        $class->error_message("Cannot create observer.  Class $class has no property ".join(',',keys %extra));
-        return;
-    }
+    return $self;
+}
 
-    my $subject_class_name = $rule->value_for('subject_class_name');
-    my $subject_class_meta = eval { $subject_class_name->__meta__ };
-    if ($@) {
-        $class->error_message("Can't create observer with subject_class_name '$subject_class_name': Can't get class metadata for class '$subject_class_name': $@");
-        return;
-    }
-    unless ($subject_class_meta) {
-        $class->error_message("Class $subject_class_name cannot be the subject class for an observer because there is no class metadata");
-        return;
-    }
 
-    my $aspect = $rule->value_for('aspect');
-    my $subject_id = $rule->value_for('subject_id');
-    unless ($subject_class_meta->_is_valid_signal($aspect)) {
-        if ($subject_class_name->can('validate_subscription') and ! $subject_class_name->validate_subscription($aspect, $subject_id, $callback)) {
-            $class->error_message("'$aspect' is not a valid aspect for class $subject_class_name");
-            return;
+{
+    my @has_defaults = qw(aspect note once priority subject_class_name subject_id);
+    sub has_defaults { @has_defaults }
+    my %defaults =
+        map {
+            $_ => __PACKAGE__->__meta__->{has}->{$_}->{default_value}
+        }
+        grep {
+            exists __PACKAGE__->__meta__->{has}->{$_}
+            && exists __PACKAGE__->__meta__->{has}->{$_}->{default_value}
+        } @has_defaults;
+    sub defaults_for_register_callback { %defaults }
+    sub register_callback {
+        my $class = shift;
+        my %params = @_;
+
+        unless (defined $params{id}) {
+            $params{id} = UR::Object::Type->autogenerate_new_object_id_uuid;
         }
-    }
 
-    if (!defined($subject_class_name) or $subject_class_name eq 'UR::Object') { $subject_class_name = '' }; # This was part of the old API, not sure why it's still here?!
-    if (!defined ($aspect)) { $aspect = '' };
-    if (!defined ($subject_id)) { $subject_id = '' };
-
-    my $self;
-    if ($method eq 'create') {
-        $self = $class->SUPER::create($rule);
-    } elsif ($method eq '__define__') {
-        $self = $class->SUPER::__define__($rule->params_list);
-    } else {
-        Carp::croak('Instantiating a UR::Observer with some method other than create() or __define__() is not supported');
-    }
-    $self->{callback} = $callback;
+        my %values = %defaults;
```

Comment:
Since we have the new `sub`s, would it be beneficial to close the block before this one and let `register_callback` call `defaults_for_register_callback` instead of using `%defaults`?